scripts/migration/migrate_ultimate_parent_node.py

```diff
@@ -0,0 +1,58 @@
+"""
+This will add an ultimate_parent field to all nodes.
+Ultimate_parent will be the primary key of the originating parent node
+"""
+
+import sys
+import logging
+from modularodm import Q
+from website import models
+from website.app import init_app
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def main():
+    init_app(routes=False)
+    dry_run = 'dry' in sys.argv
+    logger.warn('ultimate_parent field will be added to all nodes.')
+    if dry_run:
+        logger.warn('Dry_run mode')
+    all_nodes = models.Node.find()
+    all_nodes_count = all_nodes.count()
+    touched_counter = 0
+    logger.info('There are {} total nodes'.format(all_nodes_count))
+    for node in all_nodes:
+        if not getattr(node, 'parent', None):
+            touched_counter += 1
+            children = [child for child in node.get_descendants_recursive(include=lambda n: n.primary)]
+            if dry_run:
+                logger.info(
+                    '{}/{}: touched node {} with children {}'.format(
+                        touched_counter,
+                        all_nodes_count,
+                        node._id,
+                        [child._id for child in children]
+                    )
+                )
+            else:
+                node.save()
+            for child in children:
+                touched_counter += 1
+                if dry_run:
+                    logger.info(
+                        '{}/{}: touched node {} with parent {}'.format(
+                            touched_counter,
+                            all_nodes_count,
+                            child._id,
+                            child.parent_id
+                        )
+                    )
+                else:
+                    child.save()
+    assert (all_nodes_count == touched_counter)
```

Comment:
This is throwing an error with a recent staging DB — a single node isn’t being hit because its parent doesn’t think it’s its child:
``` python
In [44]: problem
Out[44]: <Node(title=u'Title for a test Node.', category=u'project') with _id u'kszcd'>

In [45]: problem.parents
Out[45]: [<Node(title=u'Private Node', category=u'project') with _id u'5xe8z'>]

In [46]: ppar = Node.load('5xe8z')

In [47]: pci = [e._id for e in ppar.get_descendants_recursive(include=lambda n: n.primary)]

In [48]: problem._id in pci
Out[48]: False

In [49]: problem.creator
Out[49]: <User(u'cwisecarver+ultimate-user@cos.io') with id u'y352a'>
```
Consult @sloria to see if this is a problem on prod, and blame @cwisecarver for causing trouble.

To get around this, you could keep track of all the `_id`s for `touched_nodes`, find the set difference above this assert (e.g. `untouched_nodes = list(set([e._id for e in all_nodes])-set(touched_nodes))`), and then `save` them, but this might not be necessary.